Observability

Psykopear opened the pull request:
This PR adds observability in Bytewax.

This does not use the python logging module, but allows for external configuration anyway.
What we gain is not having to keep the GIL when logging something, but it's not integrated with the python logging's module anymore.

By default, bytewax logs `ERROR` messages to stdout.
The logging level can be configured as an argument to the setup_tracing function.

Logs are part of the tracing infrastructure. Every log is now structured, so we can use those logs and the additional instrumentation to send "spans" and "traces" to an opentelemetry compatible backend.

This PR adds support for two backends:
- Jaeger
- Opentelemetry collector

The Jaeger backend sends data directly to Jaeger. This makes the setup simpler if that's what you are using, but it only works with Jaeger.

The Opentelemetry collector backend sends data to the opentelemetry collector. This can be configured to send traces and logs to different backends, including Jaeger, and it's the most flexible way of configuring tracing, since you can change your tracing backend without touching the dataflow, just change the collector's configuration and everything should keep working.

Psykopear commented on pysrc/bytewax/tracing.py:
```diff
@@ -0,0 +1,42 @@
+from .bytewax import (  # noqa: F401
+    TracingConfig,
+    JaegerConfig,
+    OltpTracingConfig,
```
Yep, we could hide this from the API since the configuration happens with the environment variable. Unless I end up exposing some parameter before the PR is merged, I'll remove it from here

Psykopear commented on src/inputs/kafka_input.rs:
```diff
@@ -152,7 +152,7 @@ impl KafkaInputConfig {
     let timeout = Some(Duration::from_secs(5));
 
-    log::debug!("Attempting to fetch metadata from {}", topic);
+    tracing::debug!("Attempting to fetch metadata from {}", topic);
```
That's interesting. Yesterday I realized I'm not properly propagating the traces context, so I'll have to fix this first and then we can try to integrate kafka headers information if they are there

Psykopear commented on src/operators/mod.rs:
```diff
@@ -21,23 +20,18 @@ pub(crate) mod reduce;
 pub(crate) mod stateful_map;
 
+#[tracing::instrument(level = "trace")]
 pub(crate) fn map(mapper: &TdPyCallable, item: TdPyAny) -> TdPyAny {
```
I have to try what happens with instrumentation on both sides, I'll let you know

Psykopear replied:
Just tried and nope, instrumentation works correctly, but spans used in python are not seen as children of the current span we have in Rust. Which is right I guess, but maybe after everything is fixed on the Rust side we could find a way to make python instrumentation work together with the one in Rust

Psykopear commented on docs/articles/getting-started/observability.md:
````diff
@@ -0,0 +1,117 @@
+BYTEWAX_LOG="bytewax=trace" python dataflow.py
+```
+
+The `TRACE` level is the most verbose, because it also sends traces to an opentelemetry compatible backend,
````
Nope, traces are always sent to the backend, what I wanted to note is that the `TRACE` level is super verbose, and is the one you want to use if you want proper tracing, since all the spans are at the trace level. I'll rephrase this

Psykopear commented on examples/wikistream.py:
```diff
@@ -71,4 +72,9 @@ def keep_max(max_count, new_count):
         flow,
         recovery_config=SqliteRecoveryConfig("."),
+        tracing_config=OltpTracingConfig(
+            url=os.getenv("BYTEWAX_OTLP_URL", "127.0.0.1:4317"),
```
Ups, this shouldn't have been committed, I'll create a separate example for the tracing config.
Anyway `http://127.0.0.1:4317` would work

Psykopear commented on docs/articles/getting-started/observability.md:
```diff
@@ -0,0 +1,117 @@
+        # Add this lines
+        tracing_config=OltpTracingConfig(
+            service_name="Wikistream",
+            url="http://127.0.0.1:4317",
```
`http` for the url is ok, since we force the protocol with the config option.

Psykopear replied:
I exposed the same kind of configuration that opentelemetry exposes in Rust, but I agree that having an `http` url and a `grpc` protocol might be confusing. I'm going to try if changing the url automatically chooses the correct protocol, and if that's the case we can just avoid the `protocol` arg.

Psykopear replied:
So, just changing the URL doesn't work. For example if I configure the collector to only receive `http` and set the url to `http://localhost:4318`, it still tries to connect with GRPC (which is the default protocol). I can parse the url, but I think it'd be wise to add a dependency on a crate to do that properly, and then I can fail if the scheme is not `http` or `grpc`, and force the protocol down in the collector's configuration

Psykopear commented on docs/articles/getting-started/observability.md:
````diff
@@ -0,0 +1,117 @@
+python dataflow.py
+```
+
+Open your browser at [http://127.0.0.1:16686](http://127.0.0.1:16686) and take a look at traces coming into Jaeger's UI.
````
Nope, this is an issue. As I said in a previous comment, I realized yesterday that I'm not properly propagating the trace context, so it works correctly with 1 worker and 1 process, but it starts sending spans with an invalid parent as soon as you add more workers/processes.

Psykopear replied:
After looking at it for the whole day, I'm pretty sure the problem is not the propagation of context, but the correct opening of parent spans. Context propagation should be useful to group spans from different workers under a single parent span, but that's not what's causing this issue

Psykopear replied:
Fixed

Psykopear commented on src/inputs/manual_input.rs:
```diff
@@ -104,6 +104,7 @@ impl ManualInput {
 }
 
 impl InputReader<TdPyAny> for ManualInput {
+    #[tracing::instrument(name = "manual_input", level = "trace", skip_all)]
```
This should not be a problem, I just need to add a span where we actually call the python function here. I'll try and let you know

Psykopear replied:
Added the child span. In other operator I add the name of the function to the span parameters. Here I'd have to keep around the name of the builder function and add it as a parameter just for that, but since we can only have one input I'm not sure it's worth the trouble.

Psykopear commented on src/execution/periodic_epoch.rs:
```diff
@@ -70,6 +71,56 @@ impl PeriodicEpochConfig {
+            if let Some(span) = self.epoch_span.take() {
+                span.exit();
+            }
+            self.epoch_span = Some(tracing::trace_span!("Epoch", epoch = epoch).entered());
```
Yes, should probably find a better name for this, I'm open to suggestions.

The idea of dividing the traces into epochs is to avoid having one giant span for each dataflow, but also avoid having one little span for each `step`, so I used the periodic epoch, but I agree we should not leak this information to the user.

Psykopear replied:
Yes, the epoch subdivision was a way to avoid having a single infinite span for each dataflow. The problem being that if we never close the span, the child spans are never correctly grouped.
But I agree that tying the spans to epochs might make things confusing, or as you say give wrong informations.
I found a better way of doing this thing, unrelated to epochs, but with a similar behaviour: open a new span at the beginning of the dataflow run, and close it/open a new one after a fixed (configurable) amount of time.

Psykopear commented on src/tracing/oltp_tracing.rs:
```diff
@@ -0,0 +1,127 @@
+        }
+
+        // Create the tracer
+        let tracer = opentelemetry_otlp::new_pipeline()
```
mmh, I assumed sampling would be configured in the otlp itself, but I still need to check this is supported.

Psykopear replied:
Sampling is supported by the OLTP collector itself, but it might make sense to sample data before sending to it. I added a config option, by default sampling ratio is 1 (all traces are sampled).

davidselassie commented on src/inputs/kafka_input.rs:
```diff
@@ -267,6 +267,7 @@ impl KafkaInput {
 }
 
 impl InputReader<TdPyAny> for KafkaInput {
+    #[tracing::instrument(name = "kafka_input", level = "trace", skip_all)]
```
Do these macros automatically cause the span resulting from a call to this function to be grouped under the currently open "step" and "epoch"? Or how is the hierarchy defined?

Psykopear replied:
Yes, it's basically time related, if the instrumented function is called while another span is open, it will be considered a child of that span.

davidselassie commented on src/execution/periodic_epoch.rs:
```diff
@@ -70,6 +71,56 @@ impl PeriodicEpochConfig {
 
+// Struct used to make it clear what we are doing with tracing spans,
+// for internal use of this module.
+struct TracingHelper {
```
I think it'd be useful to try out a different span hierarchy, and think of one more useful to people using dataflows. Here are three ideas that come to mind:

## 1. Data-centric

In a sense, items flowing through a dataflow are the similar to HTTP requests flowing through a microservice graph. If you're writing a dataflow and want to see what your data is doing, you'll be thinking about how to individual items flow. So do something like assign a "request ID" to each item emitted from the input so the item is the top-most, then have each operator's work on that item contribute a sub-span. End the top-level item span when all the resulting items are sent through output.

I'm not totally sure how stateful operators or windowing or aggregation would work in this context. Like if you're summing values for an hour long window, do you keep each item open until the sum output is complete? We'd then need to allow the state items to build up sets of relevant request IDs. I don't think there's any analogous problem in the microservices world as requests never coalesce?

This seems complicated to get right and revolves having a lot of metadata attached to each `TdPyAny` flowing through, but would be super cool.

One downside is you don't see operational interactions: like if one operator is hogging all the CPU time, that might be hard to know. But maybe it doesn't matter?

## 2. Epoch-centric

What I mentioned in another comment, but again: each top-level span is an epoch and you have sub-spans the part of each operator activation that works on each specific epoch. This would let you see interleaving of work for different epochs and know if there's some internal bottlenecks. But I'm not sure if it's really that helpful. It doesn't reveal anything about data dependencies.

## 3. Operational Execution-centric

This feels like a mode for us as Bytewax developers. Have each call to `Worker.step` be a top-level span, and see what operator's awakenings take up fractions of that time. We'd get a good sense of where each worker is spending CPU or wait time, but you don't learn much about the data directly as each batch is monolithic. You learn about the efficiency of relative parts and internal operator design.

I think the biggest benefit of this third option is that it requires no modifications to the dataflow itself to thread through tracing information.

Psykopear replied:
I pushed the changes I was working on on Friday. It is a mix of 3 and a different interpretation of 2.

Basically instead of using actual epochs to subdivide the spans, I added a more agnostic `periodic` span, which is closed and reopened every 10 seconds (we could make this configurable).

As I said in another comment, we could now have a proper `step` sudivision, since we handle spans where we call `worker.step`.

The problem is that since we never park the worker, we get a lot of empty `step` spans (like 10x the number of spans without it on the wikistream example). I'm going to check if I can filter out or aggregate empty step spans with a custom sampler anyway, because I find the `step` subdivision can be useful, but not if it floods the data with "useless" info.

Ideally I'd love to have `1`, a data centric subdivision, but as you noted it is non trivial as soon as you introduce stateful operators and/or multiple captures and event time processing (since spans are grouped together by system time).

We could handle that by doing something similar to what we do for recovery, and observes the data flowing through and build spans manually with the proper info for correct grouping, but this will add an overhead.

I think tracing itself should be as low cost as possible, but in the context of observability this is something I'd like to have.

I think it makes sense to have tracing itself be operational centric, but still give the users another way to observe the data flowing, which might be non zero-cost, so that if you want more advanced views of your data you can opt in knowing it doesn't come for free.

davidselassie reviewed (commented):
I think this is great work and I think the communication and config foundation looks good. But I think we need to think a little more about what useful tracing information actually looks like to someone writing a dataflow. We're also going to be fighting with Timely's execution model a lot here, as it does not provide any way to globally correlate things unless you explicitly use the epoch. Also it lets you be _very_ flexible in how you use the epoch so we might have to manually annotate our operators to note what we want.

I have a few half-baked ideas in the big comment below.

Psykopear commented on pysrc/bytewax/tracing.py:
```diff
@@ -0,0 +1,42 @@
+to offer a really detailed view of what is happening in your dataflow.
+
+By default, bytewax sends all "error" logs to the standard output.
+This can be configured with an environment variable, "BYTEWAX_LOG".
```
Currently the env var is only used for logging to console, not for configuring tracing.

I opted for this since logging is usually configurable this way, so that when a dataflow is deployed, you can change the logging level without touching the code. Thinking about it, in our case this problem is mitigated by the fact that you usually don't need to rebuild a container if you change the code, and also using something like `proc_env` would still allow users to do that through env var.

I'll work on this to make it more consistent with the rest of the config options

Psykopear replied:
Ok, I made the `log_level` a kwarg for `run_main` and `spawn_cluster` and `cluster_main`, and added the option to retrieve it in the parse module, let me know if this is what you meant.

Psykopear commented on src/execution/mod.rs:
```diff
@@ -640,13 +655,20 @@ fn worker_main<A: Allocate>(
+    recovery_config = "None",
+    tracing_config = "None"
+)]
+#[pyo3(text_signature = "(flow, *, epoch_config, recovery_config tracing_config)")]
```
fixed

Psykopear commented on src/execution/periodic_epoch.rs:
```diff
@@ -108,6 +162,14 @@ where
                 assert!(output_cap.time() == state_update_cap.time());
                 let epoch = output_cap.time();
 
+                // Close `step_span` at each step, if previously open
```
Since the change to a periodic span, I removed the `step` subdivision. We could actually introduce a proper one now, since the spans are handled in `run_until_done`, where we advance step by step, but with the "real" steps we get a lot of empty spans, since the flow spins through steps as fast as possible while waiting for things to do.

So for now we only have the periodic span, and subspans generated by the rest of the instrumentation.

davidselassie replied:
Ah right, I forgot about `Worker::step` being called "step". I am curious why we would have a lot of "empty steps" then. I thought in general workers would queue up their own work. But I guess that's a separate issue. Makes sense.

Psykopear replied:
That was in the wikistream example, where most of the time is spent waiting for new input, especially with more than one process, since we only generate the input on the first one. In those cases the worker keeps calling `step` and doing nothing since input is not received continuosly.

Psykopear commented:
With this latest commit I removed the `StdOutTracingConfig` since the only configurable parameter is the log level which can now be configured with kwargs in the execution functions, and also removed some code duplication across Tracer configs, so that we initialize the global subscriber in one place rather than in each tracing config.

Psykopear commented on .github/workflows/CI.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
       run: |
         sudo apt update
-        sudo apt install libssl-dev libsasl2-dev pkg-config openssl
+        sudo apt install libssl-dev libsasl2-dev pkg-config openssl protobuf-compiler
```
thanks! that saved me a lot of troubles :smile:

Psykopear commented:
The latest commit changes how we do the setup. Since the tracing/logging subscriber is a global that can be initialized only once per process, doing that in the execution functions is not right, and will fail if the user runs the execution function more than once (it happened in tests too).

With the latest commit, tracing has to be setup globally once from python:
```python
from bytewax.tracing import setup_tracing

tracer = setup_tracing()
```

And the `tracing_config` and `log_level` parameters can be passed to the `setup_tracing` function, so I also removed the new code in `bytewax.parse`.
